buildSrc/src/main/java/io/opentelemetry/instrumentation/gradle/muzzle/MuzzlePlugin.java

```diff
@@ -0,0 +1,518 @@
+    allRangeRequest.setRepositories(repos);
+    allRangeRequest.setArtifact(allVersionsArtifact);
+    final VersionRangeResult allRangeResult;
+    try {
+      allRangeResult = system.resolveVersionRange(session, allRangeRequest);
+    } catch (VersionRangeResolutionException e) {
+      throw new RuntimeException(e);
+    }
+
+    VersionRangeRequest rangeRequest = new VersionRangeRequest();
+    rangeRequest.setRepositories(repos);
+    rangeRequest.setArtifact(directiveArtifact);
+    final VersionRangeResult rangeResult;
+    try {
+      rangeResult = system.resolveVersionRange(session, rangeRequest);
+    } catch (VersionRangeResolutionException e) {
+      throw new RuntimeException(e);
+    }
+
+    allRangeResult.getVersions().removeAll(rangeResult.getVersions());
+
+    for (String version :
+        filterVersions(allRangeResult, muzzleDirective.getNormalizedSkipVersions())) {
+      MuzzleDirective inverseDirective =
+          instrumentationProject.getObjects().newInstance(MuzzleDirective.class);
+      inverseDirective.getGroup().set(muzzleDirective.getGroup());
+      inverseDirective.getModule().set(muzzleDirective.getModule());
+      inverseDirective.getVersions().set(version);
+      inverseDirective.getAssertPass().set(!muzzleDirective.getAssertPass().get());
+      inverseDirectives.add(inverseDirective);
+    }
+
+    return inverseDirectives;
+  }
+
+  private static Set<String> filterVersions(VersionRangeResult range, Set<String> skipVersions) {
+    Set<String> result = new HashSet<>();
+
+    AcceptableVersions predicate = new AcceptableVersions(skipVersions);
+    if (predicate.test(range.getLowestVersion())) {
+      result.add(range.getLowestVersion().toString());
+    }
+    if (predicate.test(range.getHighestVersion())) {
+      result.add(range.getHighestVersion().toString());
+    }
+
+    List<Version> copy = new ArrayList<>(range.getVersions());
+    Collections.shuffle(copy);
+    for (Version version : copy) {
+      if (result.size() >= RANGE_COUNT_LIMIT) {
+        break;
+      }
+      if (predicate.test(version)) {
+        result.add(version.toString());
+      }
+    }
+
+    return result;
+  }
+
+  /** Create muzzle's repository system */
+  private static RepositorySystem newRepositorySystem() {
+    DefaultServiceLocator locator = MavenRepositorySystemUtils.newServiceLocator();
+    locator.addService(RepositoryConnectorFactory.class, BasicRepositoryConnectorFactory.class);
+    locator.addService(TransporterFactory.class, HttpTransporterFactory.class);
+
+    return locator.getService(RepositorySystem.class);
+  }
+
+  /** Create muzzle's repository system session */
+  private static RepositorySystemSession newRepositorySystemSession(
+      RepositorySystem system, Project project) {
+    DefaultRepositorySystemSession session = MavenRepositorySystemUtils.newSession();
+    File muzzleRepo = project.file("build/muzzleRepo");
```

Comment:
I could use a normal temp directory life before but figured it's nice to use user temp directory as little as possible for Gradle builds